profiler/internal/stackparse/stackparse.go

```diff
@@ -0,0 +1,340 @@
+	// Multiple matched pairs are allowed, but nesting is not.
+	var (
+		openIndex  = -1
+		closeIndex = -1
+	)
+	for i, r := range line {
+		switch r {
+		case '(':
+			if openIndex != -1 && closeIndex == -1 {
+				return nil
+			}
+			openIndex = i
+			closeIndex = -1
+		case ')':
+			if openIndex == -1 || closeIndex != -1 {
+				return nil
+			}
+			closeIndex = i
+		}
+	}
+
+	if openIndex == -1 || closeIndex == -1 || openIndex == 0 {
+		return nil
+	}
+	return &Frame{Func: string(line[0:openIndex])}
+}
+
+// parseFile parses a file line and updates f accordingly or returns false on
+// error.
+//
+// Example Input:
+// /root/go1.15.6.linux.amd64/src/net/http/server.go:2969 +0x36c
+//
+// Example Update:
+// &Frame{File: "/root/go1.15.6.linux.amd64/src/net/http/server.go", Line: 2969}
+func parseFile(line []byte, f *Frame) bool {
+	if len(line) == 0 || line[0] != '\t' {
+		return false
+	}
+
+	line = line[1:]
+	for i, c := range line {
+		if c == ':' {
+			if f.File != "" {
+				return false
+			}
+			f.File = string(line[0:i])
+		} else if c == ' ' || i+1 == len(line) {
+			if f.File == "" {
+				return false
+			}
+			var end int
+			if c == ' ' {
+				end = i
+			} else {
+				end = i + 1
+			}
+
+			var err error
+			f.Line, err = strconv.Atoi(string(line[len(f.File)+1 : end]))
+			return err == nil
+		}
+	}
+	return false
+}
+
+// Goroutine represents a single goroutine and its stack after extracting it
+// from the runtime.Stack() text format. See [1] for more info.
+// [1] https://github.com/felixge/go-profiler-notes/blob/main/goroutine.md
+type Goroutine struct {
+	// ID is the goroutine id (aka `goid`).
+	ID int
+	// State is the `atomicstatus` of the goroutine, or if "waiting" the
+	// `waitreason`.
+	State string
+	// Wait is the approximate duration a goroutine has been waiting or in a
+	// syscall as determined by the first gc after the wait started. Aka
+	// `waitsince`.
+	Wait time.Duration
+	// LockedToThread is true if the goroutine is locked by a thread, aka
+	// `lockedm`.
+	LockedToThread bool
+	// Stack is the stack trace of the goroutine.
+	Stack []*Frame
+	// FramesElided is true if the stack trace contains a message indicating that
+	// additional frames were elided. This happens when the stack depth exceeds
+	// 100.
+	FramesElided bool
+	// CreatedBy is the frame that created this goroutine, nil for main().
+	CreatedBy *Frame
+}
+
+// Frame is a single call frame on the stack.
+type Frame struct {
+	// Func is the name of the function, including package name, e.g. "main.main"
+	// or "net/http.(*Server).Serve".
+	Func string
+	// File is the absolute path of source file e.g.
+	// "/go/src/example.org/example/main.go".
+	File string
+	// Line is the line number of inside of the source file that was active when
+	// the sample was taken.
+	Line int
+}
+
+// Errors contains a list of errors.
+type Errors struct {
+	// Errors is a list of errors.
+	Errors []error
+}
```

Comment:
I've gone back and forth on whether we need a type for this. Maybe it's overkill.